### DISCUSSION TITLE
Collection takes too long if I don't specify each subdirectory

### DISCUSSION THREAD

**Discussion by joshibeast** [pytest-dev/pytest#11004] [Q&A] [answered]
Hey folks,
I'm just starting to use pytest to run some regression security tests but I have a problem that I don't know how to solve.

Here is where I'm running the tests from:

    
<img width="354" alt="Screenshot 2023-05-15 at 08 11 49" src="https://github.com/pytest-dev/pytest/assets/10061471/302cae0f-c447-4b03-875e-2a7ed8791235">

There are total of 300 tests within the test directories.

When I run the following command:
pytest tests/* - collection takes ~20 minutes

However, if I try to run:
pytest tests/tests_dir1 tests/tests_dir2 tests/tests_dir3 .... tests/tests_dir10 - specifying each subdirectory where the tests are it takes only couple of minutes.

Tests are being triggered by gitlab runner, I mean I have an option to dynamically generate the second command during each pipeline trigger but it just seems like a wrong approach here because I'm obviously doing something wrong in the first place.

Thanks for your help.

**Comment by RonnyPfannschmidt** [comment]
There is something confusing there

Shell expansion ought to imply that both the explicit list and the glob are the same

What's the output of echo `tests/*`

**Reply by RonnyPfannschmidt** [reply]
Also please provide extra details around python version, pytest version and used images

**Reply by joshibeast** [reply]
Python 3.10.2
pytest 7.2.1

echo tests/* = just these test dirs + __pycache__
I also have pycache dirs across all subdirectories in test/ but I don't think they are my problem.

**Reply by RonnyPfannschmidt** [reply, marked as answer]
Indeed, that "should" not to be a problem

What's the difference in output when the PYTEST_DEBUG is set and --collectonly is passed?

**Reply by joshibeast** [reply]
Thanks! This solved it for me and I feel a bit dull at the moment.
There was an unrelated script/test named something.py in tests/ root that executed a very long ~20 minute job.
That's where and why collection hanged.
I wasn't aware that pytest will actually try to run this script because it doesn't contain test_ in filename.

**Reply by RonnyPfannschmidt** [reply]
The filename was explicitly passed, so it tries, always use main functions and the dunder name check to prevent import runaways

**Reply by RonnyPfannschmidt** [reply]
Also don't fret it, it's a very classical human error most of us hit one way or another, at least Bobby tables didn't enrol
